Initial framework for HybridAutomaton

jadecastro opened the pull request:
Adds a ModalSubsystem class to an otherwise empty HybridAutomatonContext. A follow-on PR will introduce the rest of HybridAutomatonContext.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5083)
<!-- Reviewable:end -->

jadecastro commented:
+@david-german-tri for feature review, please. (I'll defer platform review, since this is lower priority)

david-german-tri commented:
+(status: curate commits before merging) 

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-:-KcTqxwHrbYspg1b-6Bo:b-s794vx)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
First pass.

---

Reviewed 5 of 6 files at r1.
Review status: 5 of 6 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/systems/framework/hybrid_automaton_context.cc, line 1 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTrn1d4rlZs3kPmF_7:-KcTrn1eKVJWLeQAKdgV:bmo30y6) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.cc#L1)):*
> ```cpp
> #include "drake/systems/framework/hybrid_automaton_context.h"
> ```

Please include explicit instantiations - at least double, and possibly autodiff or symbolic if you want to add tests for them.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTsF-3oX7-f45gu3vL:-KcTsF-4i9faPdDjzzY4:bcdu8c4) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L22)):*
> ```objc
> using std::unique_ptr;
> 
> /*
> ```

BTW should this be doxygen?  If so, please make sure it compiles into something reasonable.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 55 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTs4siao6XXL-i43OK:-KcTs4siao6XXL-i43OL:b-t3j0wn) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L55)):*
> ```objc
> */
> template <typename T>
> class ModalSubsystem {
> ```

The filename should match the class name.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTumh3v1Xh9wedtRo4:-KcTumh3v1Xh9wedtRo5:be0cwe9) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L62)):*
> ```objc
>   typedef int PortId;
> 
>   // Constructor
> ```

Please document the parameters.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcU1AJSuejO3nOo7SCp:-KcU1AJSuejO3nOo7SCq:bbniisb) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L72)):*
> ```objc
>         invariant_(invariant),
>         initial_conditions_(initial_conditions),
>         input_port_ids_(input_port_ids),
> ```

Should these be validity-checked?

---

*[drake/systems/framework/hybrid_automaton_context.h, line 77 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTv4f9MrswCk2sJg3q:-KcTv4f9MrswCk2sJg3r:b8ysur9) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L77)):*
> ```objc
>   }
> 
>   ModalSubsystem(ModeId mode_id, shared_ptr<System<T>> system,
> ```

This constructor and the one below should delegate to the more general constructor.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTv9kO8F6jCN--9EhV:-KcTv9kO8F6jCN--9EhW:byk721z) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L95)):*
> ```objc
>   /// Accessors for the underlying data.
>   ModeId get_mode_id() const { return mode_id_; }
>   System<T>* get_system() const { return system_.get(); }
> ```

I'm uneasy that a `const` method returns a non-const pointer.  Is that necessary?  If so, please add a comment justifying it.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 96 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcU0AIQOHIkrpPB_RYw:-KcU0AIQOHIkrpPB_RYx:b-wnqbyt) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L96)):*
> ```objc
>   ModeId get_mode_id() const { return mode_id_; }
>   System<T>* get_system() const { return system_.get(); }
>   const std::vector<PortId> get_input_port_ids() const {
> ```

These get input/output port ID methods need some kind of comment explaining the semantics (which I think is a remapping?)

---

*[drake/systems/framework/hybrid_automaton_context.h, line 102 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTvj5nC36rBrmZFv3O:-KcTvj5nC36rBrmZFv3P:b-5vetc5) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L102)):*
> ```objc
>   PortId get_input_port_id(const int index) const {
>     DRAKE_DEMAND(index >= 0 &&
>                  index < static_cast<int>(input_port_ids_.size()));
> ```

BTW `get_num_input_ports()`?

---

*[drake/systems/framework/hybrid_automaton_context.h, line 125 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTvyAmwsjLNp40b5SA:-KcTvyAmwsjLNp40b5SB:b-cm0s8d) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L125)):*
> ```objc
>   /// Accessors for the symbolic::Expressions for the invariants and initial
>   /// condition state sets for this ModalSubsystem. Their defining sets are
>   /// semialgebraic: a state assignment is within the set iff it evaluates to a
> ```

BTW would it be possible to put some example code in the comment to show what you mean here?

---

*[drake/systems/framework/hybrid_automaton_context.h, line 180 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTwUzSGg2KbBIewhWc:-KcTwUzSGg2KbBIewhWd:b-3bhfpr) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L180)):*
> ```objc
> 
>   // Creates symbolic variables according to the variable_type key word.
>   void CreateSymbolicVariables(const std::string variable_type,
> ```

BTW  `const std::string&`

---

*[drake/systems/framework/hybrid_automaton_context.h, line 191 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTwq7unIlayutaPUmK:-KcTwq7unIlayutaPUmL:b-uw889r) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L191)):*
> ```objc
>       key << variable_type << i;
>       symbolic::Variable var{key.str()};
>       row.emplace_back(var);
> ```

BTW you can collapse these two lines into `row.emplace_back(key.str())`

---

*[drake/systems/framework/hybrid_automaton_context.h, line 209 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTuwMY3mWB1bCn3XAY:-KcTuwMY3mWB1bCn3XAZ:b-2augz) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L209)):*
> ```objc
>   // TODO(jadecastro): Allow ModeId to take on a descriptor in place of the int.
>   ModeId mode_id_;
>   // The system model.
> ```

Please explain why this is a `shared_ptr`.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 218 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTx-LUgi-zvgG6RBzK:-KcTx-LUgi-zvgG6RBzL:b-fpmsdf) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L218)):*
> ```objc
>   std::vector<PortId> input_port_ids_;
>   std::vector<PortId> output_port_ids_;
>   // A vector of symbolic variables for each of the continuous states in the
> ```

Is this comment correct?  (a) it doesn't match the shape of the type, and (b) it seems like you're using it for discrete states also.

---

*[drake/systems/framework/test/hybrid_automaton_context_test.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTy0V_He7RMjBlOBNU:-KcTy0V_He7RMjBlOBNV:bntknun) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/test/hybrid_automaton_context_test.cc#L34)):*
> ```cpp
> constexpr double kZohSamplingPeriod{0.1};
> 
> class AbstractTestSource : public LeafSystem<double> {
> ```

Why not just use `ConstantValueSource`?  It's not the same thing, but if anything I think it broadens the coverage of this test.

---

*[drake/systems/framework/test/hybrid_automaton_context_test.cc, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTyMnKlVmMESAhVCMi:-KcTyMnKlVmMESAhVCMj:b-d2nt6k) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/test/hybrid_automaton_context_test.cc#L88)):*
> ```cpp
>     integrator_->set_name("Alice");
> 
>     // Instantiate a new modal subsystem. Implicitly, one input and one output
> ```

I don't understand the second sentence.

---

*[drake/systems/framework/test/hybrid_automaton_context_test.cc, line 99 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcU-NVvRFa65qEGifr-:-KcU-NVvRFa65qEGifr0:br44ggs) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/test/hybrid_automaton_context_test.cc#L99)):*
> ```cpp
> };
> 
> class HybridAutomatonContextStateTest : public ::testing::Test {
> ```

BTW please sort these so the fixture immediately precedes its `TEST_F` cases.

---

*[drake/systems/framework/test/hybrid_automaton_context_test.cc, line 125 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcU-bgt3GbBGaaMFc9F:-KcU-bgt3GbBGaaMFc9G:bdogb3r) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/test/hybrid_automaton_context_test.cc#L125)):*
> ```cpp
>   EXPECT_EQ(0, mss_->get_output_port_ids()[0]);
> 
>   // Explicitly specify the ports for a non-trivial example via the constructor.
> ```

Please split this into two separate cases, one of which doesn't need the fixture at all.

---

*[drake/systems/framework/test/hybrid_automaton_context_test.cc, line 167 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcU22neevS6ew5J78PX:-KcU22neevS6ew5J78PY:b-1ujqvg) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/test/hybrid_automaton_context_test.cc#L167)):*
> ```cpp
>   // Evaluate the expressions for a given assignment.
>   symbolic::Environment x_env{{xc, 3.}, {xd, -2.}};
>   EXPECT_EQ(pow(3., 2.) - cos(3.) + tanh(-2.),
> ```

These appear to be assertions on the symbolic library, not on `ModalSubsystem`.  Do we really need them?  Can you instead write a test that demonstrates actual usage of invariants/initial conditions?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-:-KcU2HJsarYGAIZ40zDL:b-xepgfs)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 1 of 6 files at r1.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5083)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot linux-xenial-gcc-experimental-open-source please
@drake-jenkins-bot linux-xenial-cpplint-experimental please

jadecastro commented:
Sorry for the delay- I've finally just circled back around on your comments, so PTAL. Note the question on the file name/class name comment.

---

Review status: 0 of 6 files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/systems/framework/hybrid_automaton_context.cc, line 1 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTrn1d4rlZs3kPmF_7:-KdJ41ktj1k2guaQ71p9:bhyvt7m) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.cc#L1)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Please include explicit instantiations - at least double, and possibly autodiff or symbolic if you want to add tests for them.
</blockquote></details>

Done.  I've just included `<double>`, since this is the only unit test coverage I've managed to create so far.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTsF-3oX7-f45gu3vL:-Kd1vgpYHEB0Q8NIYpXi:blqyr70) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L22)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

BTW should this be doxygen?  If so, please make sure it compiles into something reasonable.
</blockquote></details>

Thanks... Done.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 55 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTs4siao6XXL-i43OK:-KcYenUpJ8WH17RbzbbL:bpqtq6b) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L55)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

The filename should match the class name.
</blockquote></details>

Note that the only reason for removing HybridAutomatonContext from this file (justified in the TODO at the bottom), was to make this PR small enough LOC limits. Do you suggest I rename the file now, then rename it again later? Or should I now plan to place each class into their own sets of files?
It seems to make the most sense to me having both ModalSubsystem and HybridAutomatonContext ultimately grouped together (similar to system_output.h, value.h).

---

*[drake/systems/framework/hybrid_automaton_context.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTumh3v1Xh9wedtRo4:-Kd0p9zajwp3_3wpbqv_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L62)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Please document the parameters.
</blockquote></details>

Done.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 72 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcU1AJSuejO3nOo7SCp:-KdOXieNOf14xOQyVT2R:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L72)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Should these be validity-checked?
</blockquote></details>

Done.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 77 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTv4f9MrswCk2sJg3q:-Kd0p6Z3GUeulBe1H8ba:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L77)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

This constructor and the one below should delegate to the more general constructor.
</blockquote></details>

Done.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTv9kO8F6jCN--9EhV:-KdOZLV2HJ3Yq-jvyc7W:b-297fdh) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L95)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

I'm uneasy that a `const` method returns a non-const pointer.  Is that necessary?  If so, please add a comment justifying it.
</blockquote></details>

Done. It's `const` now.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 96 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcU0AIQOHIkrpPB_RYw:-KdS-Ww9wAt8sXPciRbl:b37wgz1) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L96)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

These get input/output port ID methods need some kind of comment explaining the semantics (which I think is a remapping?)
</blockquote></details>

Done. I've included a more precise description in the member field comments.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 102 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTvj5nC36rBrmZFv3O:-KdOu2pwUfZl809avLTM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L102)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

BTW `get_num_input_ports()`?
</blockquote></details>

Done.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 125 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTvyAmwsjLNp40b5SA:-KdS8970rBaha6UGUlFK:b2hvdsx) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L125)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

BTW would it be possible to put some example code in the comment to show what you mean here?
</blockquote></details>

Done. I've included that in the class description, and have rearranged this text a bit with the member field comments.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 180 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTwUzSGg2KbBIewhWc:-KdOuDDSTDavyGZEOFth:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L180)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

BTW  `const std::string&`
</blockquote></details>

Done.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 191 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTwq7unIlayutaPUmK:-KdOwEHf_GqfiAO1fHRY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L191)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

BTW you can collapse these two lines into `row.emplace_back(key.str())`
</blockquote></details>

Done.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 209 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTuwMY3mWB1bCn3XAY:-KdOwzMLCeF5-QhKMrN8:btd84zi) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L209)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Please explain why this is a `shared_ptr`.
</blockquote></details>

Ultimately, I want to have HybridAutomaton retain a list of all the modes in the automaton, yet allow the context to have access to a "clone" of the _active_ ModalSubsystem as it executes so that the context can do things like check invariants and report the right mode ID to callers.  

When a HybridAutomaton-build is finalized, a new HybridAutomaton object will be populated with all of the ModalSubsystems required to properly represent the hybrid system. At runtime, one of these ModalSubsystems will be activated, and its "clone" will be shared with the HAContext.  While I am calling this "clone" (quotes intended for lack of a better term!), under the hood I am sharing resources (via `std::shared_ptr`) of System objects, so as to avoid making deep copies of Systems themselves.  Of course, whenever a transition is made, I have to take care to destroy the existing ModalSubsystem each a new one is activated.

I had given some thought to using `std::move` to the ModalSubsystem from the list retained in HybridAutomaton at every discrete event, and then `std::move` it back into the list upon seeing a new event, but that would involve some rework and I see that possibly adding further complications (like when a self-transition happens).  I'm definitely open to suggestions, though, if there's a better design choice than this.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 218 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTx-LUgi-zvgG6RBzK:-KdS12dWhSLEE9eYeL5C:b-6fjcjt) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L218)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Is this comment correct?  (a) it doesn't match the shape of the type, and (b) it seems like you're using it for discrete states also.
</blockquote></details>

Done. (I failed to update this comment after having changed the data structure itself.)

---

*[drake/systems/framework/test/hybrid_automaton_context_test.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTy0V_He7RMjBlOBNU:-KdRtLCgp5isuXDHcmEv:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/test/hybrid_automaton_context_test.cc#L34)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Why not just use `ConstantValueSource`?  It's not the same thing, but if anything I think it broadens the coverage of this test.
</blockquote></details>

Done.

---

*[drake/systems/framework/test/hybrid_automaton_context_test.cc, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTyMnKlVmMESAhVCMi:-KdRnfij4jq4n4Zt3CTl:b-qsvnoa) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/test/hybrid_automaton_context_test.cc#L88)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

I don't understand the second sentence.
</blockquote></details>

Done. PTAL.

---

*[drake/systems/framework/test/hybrid_automaton_context_test.cc, line 99 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcU-NVvRFa65qEGifr-:-KdQWjjoCaajuNDTbVdD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/test/hybrid_automaton_context_test.cc#L99)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

BTW please sort these so the fixture immediately precedes its `TEST_F` cases.
</blockquote></details>

Done.

---

*[drake/systems/framework/test/hybrid_automaton_context_test.cc, line 125 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcU-bgt3GbBGaaMFc9F:-KdROfark8rdOCxOfdQ5:bdlm9md) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/test/hybrid_automaton_context_test.cc#L125)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Please split this into two separate cases, one of which doesn't need the fixture at all.
</blockquote></details>

Done. (I've also added an additional one to test the port-validity checks.)

---

*[drake/systems/framework/test/hybrid_automaton_context_test.cc, line 167 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcU22neevS6ew5J78PX:-KdRS_01kzuBaQRrIJYX:byxlglo) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/test/hybrid_automaton_context_test.cc#L167)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

These appear to be assertions on the symbolic library, not on `ModalSubsystem`.  Do we really need them?  Can you instead write a test that demonstrates actual usage of invariants/initial conditions?
</blockquote></details>

Agree. I've changed the `EXPECT_EQ()`'s to `EXPECT_NO_THROW()` just to illustrate the fact that that any of the example expressions are valid (consistent wrt the discrete/continuous states).  Also, I've changed the comments to be more clear about the workflow on the part of the user.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-:-Kcyf1chJkSPjUTIwgpU:btco57s)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 0 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 55 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTs4siao6XXL-i43OK:-KdYDb8bC7Enp5mjFWdc:b-ix1y93) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L55)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Note that the only reason for removing HybridAutomatonContext from this file (justified in the TODO at the bottom), was to make this PR small enough LOC limits. Do you suggest I rename the file now, then rename it again later? Or should I now plan to place each class into their own sets of files?
It seems to make the most sense to me having both ModalSubsystem and HybridAutomatonContext ultimately grouped together (similar to system_output.h, value.h).
</blockquote></details>

Given that this file is already pushing 300 lines, I'd suggest separate files.  (Yes, the same probably goes for at least `system_output.h`.  Thanks for pointing that out!)

---

*[drake/systems/framework/hybrid_automaton_context.h, line 209 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KcTuwMY3mWB1bCn3XAY:-KdYHmaqRVtAVTjvwdg2:b-tqarc1) ([raw file](https://github.com/robotlocomotion/drake/blob/5976734d17aee0c7ef56fd469c589201a39efca1/drake/systems/framework/hybrid_automaton_context.h#L209)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Ultimately, I want to have HybridAutomaton retain a list of all the modes in the automaton, yet allow the context to have access to a "clone" of the _active_ ModalSubsystem as it executes so that the context can do things like check invariants and report the right mode ID to callers.  

When a HybridAutomaton-build is finalized, a new HybridAutomaton object will be populated with all of the ModalSubsystems required to properly represent the hybrid system. At runtime, one of these ModalSubsystems will be activated, and its "clone" will be shared with the HAContext.  While I am calling this "clone" (quotes intended for lack of a better term!), under the hood I am sharing resources (via `std::shared_ptr`) of System objects, so as to avoid making deep copies of Systems themselves.  Of course, whenever a transition is made, I have to take care to destroy the existing ModalSubsystem each a new one is activated.

I had given some thought to using `std::move` to the ModalSubsystem from the list retained in HybridAutomaton at every discrete event, and then `std::move` it back into the list upon seeing a new event, but that would involve some rework and I see that possibly adding further complications (like when a self-transition happens).  I'm definitely open to suggestions, though, if there's a better design choice than this.
</blockquote></details>

I don't think I understand (based on what's in the design doc) why HAContext needs access to the active `System` at all.  If we grant that it does, I also don't understand why that access would need to be `shared_ptr`, and not just `const&`.

---

*[drake/systems/framework/hybrid_automaton_context.h, line 56 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KdYIbxfmQIYv1b6jcbK:-KdYIbxfmQIYv1b6jcbL:bd8218q) ([raw file](https://github.com/robotlocomotion/drake/blob/afc4230c23512e97f9fede2395478c25ce9985d2/drake/systems/framework/hybrid_automaton_context.h#L56)):*
> ```objc
>  *
>  * The invariants and initial conditions for this ModalSubsystem are of
>  * symbolic::Expression type, with each defining a semialgebraic set: a state
> ```

Based on the following, I can at least guess why invariants are semialgebraic sets.  Initial conditions, I still have no idea.  

---

*[drake/systems/framework/hybrid_automaton_context.h, line 67 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KdYIFX9ld7fkhFh0DL3:-KdYIFX9ld7fkhFh0DL4:bow18c6) ([raw file](https://github.com/robotlocomotion/drake/blob/afc4230c23512e97f9fede2395478c25ce9985d2/drake/systems/framework/hybrid_automaton_context.h#L67)):*
> ```objc
>  *  const symbolic::Expression x0{xcs[0]};  // Symbol for the continuous state.
>  *  (*invariant).push_back({-x0 + 1.})  // The unbounded set x <= 1 defines this
>  *                                      // invariant.
> ```

This is helpful, but it's still not clear from the doc what the invariant _means_.  If any invariant is false, do we leave the mode?

---

*[drake/systems/framework/hybrid_automaton_context.h, line 248 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KdYGuK7cuZcxm_vTyFx:-KdYGuK7cuZcxm_vTyFy:bw01e16) ([raw file](https://github.com/robotlocomotion/drake/blob/afc4230c23512e97f9fede2395478c25ce9985d2/drake/systems/framework/hybrid_automaton_context.h#L248)):*
> ```objc
>   // and symbolic_variables_ will be enforced within HybridAutomatonContext.
>   // (See the class docstring for details.)
>   std::vector<symbolic::Expression> invariant_;
> ```

In the design doc, the invariants and initial conditions are `symbolic::Formula`.  Is there now an implicit ` = 0`?

---

*[drake/systems/framework/hybrid_automaton_context.h, line 257 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KdYH2wkNrkezN2jJ6uo:-KdYH2wkNrkezN2jJ6up:bkpm3yt) ([raw file](https://github.com/robotlocomotion/drake/blob/afc4230c23512e97f9fede2395478c25ce9985d2/drake/systems/framework/hybrid_automaton_context.h#L257)):*
> ```objc
>   std::vector<symbolic::Expression> initial_conditions_;
> 
>   // Index sets for the input and output ports for this mode.  Each @p PortId
> ```

BTW `PortId` isn't a parameter, so no `@p`, right?

---

*[drake/systems/framework/hybrid_automaton_context.h, line 266 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5083#-KdYEk6-HLqjolbPNFhL:-KdYEk60yeyAzMLtoD8p:b-yeo8f0) ([raw file](https://github.com/robotlocomotion/drake/blob/afc4230c23512e97f9fede2395478c25ce9985d2/drake/systems/framework/hybrid_automaton_context.h#L266)):*
> ```objc
>   // A data structure containing the symbolic variables corresponding to the
>   // states of the System.  Continuous and discrete states correspond,
>   // respectively, to the std::string keys "xc" and "xd".  For convenience,
> ```

Why not just have separate member variables for xc (`vector`) and xd (`vector<vector>`), if those are the only keys?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5083)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
I'll reopen this PR once I have addressed the comments and have given the code a much-needed update.
